source/code/include/playfab/PlayFabPlatformUtils.h

```diff
@@ -7,53 +7,127 @@
+    inline time_t TimePointToTimeT(TimePoint input)
     {
-        return std::chrono::system_clock::now();
+        return Clock::to_time_t(input);
     }
 
-    inline time_t GetPlayFabTimeTNow()
+    inline TimePoint TimeTToTimePoint(time_t input)
     {
-        return time(0);
+        return Clock::from_time_t(input);
     }
 
-    inline std::string LocalTimeTToUtcString(time_t now)
+    inline tm TimeTToUtcTm(time_t input)
     {
-        tm timeInfo;
-#if defined(PLAYFAB_PLATFORM_WINDOWS) || defined(PLAYFAB_PLATFORM_XBOX)
-        gmtime_s(&timeInfo, &now);
-#elif defined(PLAYFAB_PLATFORM_LINUX) || defined(PLAYFAB_PLATFORM_IOS) || defined(PLAYFAB_PLATFORM_ANDROID) || defined(PLAYFAB_PLATFORM_PLAYSTATION)
-        timeInfo = *gmtime(&now);
+        tm timeInfo{ 0 };
+#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
+        gmtime_s(&input, &timeInfo);
+#elif defined(PLAYFAB_PLATFORM_WINDOWS) || defined(PLAYFAB_PLATFORM_XBOX)
+        gmtime_s(&timeInfo, &input);
+#else
+        gmtime_r(&input, &timeInfo);
 #endif
-        char buff[64];
-        strftime(buff, 64, "%Y-%m-%dT%H:%M:%S.000Z", &timeInfo);
+        return timeInfo;
+    }
 
-        return buff;
+    inline time_t UtcTmToTimeT(tm input)
+    {
+#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
+        return mktime(&input);
+#elif defined(PLAYFAB_PLATFORM_WINDOWS) || defined(PLAYFAB_PLATFORM_XBOX)
+        return _mkgmtime(&input);
+#else
+        return timegm(&input);
+#endif
+    }
+
+    inline tm TimePointToUtcTm(TimePoint input)
+    {
+        return TimeTToUtcTm(Clock::to_time_t(input));
     }
 
-    inline std::string LocalTimePointToUtcString(TimePoint now)
+    inline TimePoint UtcTmToTimePoint(const tm& input)
     {
-        return LocalTimeTToUtcString(std::chrono::system_clock::to_time_t(now));
+        return TimeTToTimePoint(UtcTmToTimeT(input));
     }
 
-    inline time_t UtcStringToLocalTimeT(const std::string& utcString)
+    // Get Time now - Platform dependent granularity (granularity: upto 1 second, accuracy within a few seconds)
+    inline TimePoint GetTimePointNow()
     {
-        time_t output;
-        tm timeStruct = {};
+        // The conversion is mostly to ensure consistent behavior among all platforms
+        return std::chrono::time_point_cast<std::chrono::seconds>(Clock::now());
+    }
 
-        std::istringstream iss(utcString);
-        iss >> std::get_time(&timeStruct, "%Y-%m-%dT%T");
-        timeStruct.tm_isdst = 0;  // 0 means "not in DST" PlayFab assumes UTC/Zulu always
-#if defined(PLAYFAB_PLATFORM_PLAYSTATION)
-        output = mktime(&timeStruct);
-#elif defined(PLAYFAB_PLATFORM_IOS) || defined(PLAYFAB_PLATFORM_ANDROID) || defined(PLAYFAB_PLATFORM_LINUX)
-        gmtime_r(&output, &timeStruct);
+    inline time_t GetTimeTNow()
+    {
+        return TimePointToTimeT(GetTimePointNow());
+    }
+
+    // Get a tick count that represents now in milliseconds (not useful for absolute time)
+    inline long GetMilliTicks()
```

Comment:
recommend using uint64_t instead of long
1. fixed-width int type to make it clear what bounds the return value has (it's platform specific how big a long is)
2. the count of ticks should never be negative so returning a signed type doesn't make sense
3. if long is a 32-bit sized value it will overflow at least every 2 months (ish). This may not seem like a big deal but considering games can be suspended for LONG periods of time and resumed this is probably just better not left to chance

Reply:
If I were going to use an alternate type, then I'd probably switch to UInt64, just to be consistent with the rest of the SDK (though I know you dislike that type).
I used a long, because long is the literal exact type that matches the underlying type for all the Clock implementations (though, you have to F12 into the std headers to find it).
Only one platform needed the static_cast<long> below, and that's only because it had typedef __rep long, rather than using long directly.
So, I understand your point, but if I choose to match the underlying implementation of the clocks, I should go with long, and if I want to consistently match the established types in PlayFab, I should use the UInt64 typedef.

---
In reply to: [357800254](https://github.com/PlayFab/XPlatCoreTemplate/pull/65#discussion_r357800254) [](ancestors = 357800254)

Reply:
then I would prefer Uint64 or Int64
from looking at docs, std::chrono::milliseconds (which is the duration time you are using) says that it's underlying representation is a signed integer type of at least 45 bits. I think using Int64 or Uint64 is just a better way to communicate that [https://en.cppreference.com/w/cpp/chrono/duration](https://en.cppreference.com/w/cpp/chrono/duration)

---
In reply to: [357822540](https://github.com/PlayFab/XPlatCoreTemplate/pull/65#discussion_r357822540) [](ancestors = 357822540,357800254)